File: client/src/test/scala/stasis/test/specs/unit/client/encryption/secrets/UserKeyStoreEncryptionSecretSpec.scala
package stasis.test.specs.unit.client.encryption.secrets

import org.apache.pekko.actor.ActorSystem
import stasis.client.encryption.secrets.{DeviceSecret, UserKeyStoreEncryptionSecret}
import stasis.test.specs.unit.AsyncUnitSpec
import stasis.test.specs.unit.client.EncodingHelpers

class UserKeyStoreEncryptionSecretSpec extends AsyncUnitSpec with SecretsConfig with EncodingHelpers {
  "A UserKeyStoreEncryptionSecret" should "encrypt device secrets" in {
    keyStoreEncryptionSecret
      .encryptDeviceSecret(
        secret = deviceSecret
      )
      .map { actualEncryptedSecret =>
        actualEncryptedSecret should be(encryptedDeviceSecret)
      }
  }

  it should "decrypt device secrets" in {
    keyStoreEncryptionSecret
      .decryptDeviceSecret(
        device = testDevice,
        encryptedSecret = encryptedDeviceSecret
      )
      .map { actualDeviceSecret =>
        actualDeviceSecret should be(deviceSecret)
      }
  }

  it should "not render its content via toString" in {
    keyStoreEncryptionSecret.toString should be(s"Secret(${keyStoreEncryptionSecret.getClass.getName})")
  }

  private implicit val system: ActorSystem = ActorSystem(name = "UserKeyStoreEncryptionSecretSpec")

  private val encryptionIv = "6mSjkDoXUNNK8TGbebRYWXnjfskeVHXhaMxBRKD+ITvMckUp0ZQtUeEttz9pA0vWQ4MKa8otGmyD\nJ7OCrdeY4g=="
  private val encryptionKey = "kROAgx70MxRKeDODRMyshRH0tswcd4jydKA60r+5knI="
  private val secret = "BOunLSLKxVbluhDSPZ/wWw=="

  private val keyStoreEncryptionSecret = UserKeyStoreEncryptionSecret(
    user = testUser,
    iv = encryptionIv.decodeFromBase64,
    key = encryptionKey.decodeFromBase64
  )

  private val deviceSecret = DeviceSecret(
    user = testUser,
    device = testDevice,
    secret = secret.decodeFromBase64
  )

  private val encryptedDeviceSecret = "lEc6SelbtpNPtzcm6AjkXQR51FNeWuM0xkWSXKARiFQ=".decodeFromBase64
}

File: client-android/lib/src/test/kotlin/stasis/test/client_android/lib/encryption/secrets/UserKeyStoreEncryptionSecretSpec.kt
package stasis.test.client_android.lib.encryption.secrets

import io.kotest.core.spec.style.WordSpec
import io.kotest.matchers.shouldBe
import okio.ByteString.Companion.decodeBase64
import stasis.client_android.lib.encryption.secrets.DeviceSecret
import stasis.client_android.lib.encryption.secrets.UserKeyStoreEncryptionSecret

class UserKeyStoreEncryptionSecretSpec : WordSpec({
    "A UserKeyStoreEncryptionSecret" should {
        val encryptionIv = "6mSjkDoXUNNK8TGbebRYWXnjfskeVHXhaMxBRKD+ITvMckUp0ZQtUeEttz9pA0vWQ4MKa8otGmyD\nJ7OCrdeY4g=="
        val encryptionKey = "kROAgx70MxRKeDODRMyshRH0tswcd4jydKA60r+5knI="
        val secret = "BOunLSLKxVbluhDSPZ/wWw=="

        val keyStoreEncryptionSecret = UserKeyStoreEncryptionSecret(
            user = SecretsConfig.testUser,
            iv = encryptionIv.decodeBase64()!!,
            key = encryptionKey.decodeBase64()!!,
            target = SecretsConfig.testConfig
        )

        val deviceSecret = DeviceSecret(
            user = SecretsConfig.testUser,
            device = SecretsConfig.testDevice,
            secret = secret.decodeBase64()!!,
            target = SecretsConfig.testConfig
        )

        val encryptedDeviceSecret = "lEc6SelbtpNPtzcm6AjkXQR51FNeWuM0xkWSXKARiFQ=".decodeBase64()!!

        "encrypt device secrets" {
            val actualEncryptedSecret = keyStoreEncryptionSecret.encryptDeviceSecret(
                secret = deviceSecret
            )

            actualEncryptedSecret shouldBe (encryptedDeviceSecret)
        }

        "decrypt device secrets" {
            val actualDeviceSecret = keyStoreEncryptionSecret.decryptDeviceSecret(
                device = SecretsConfig.testDevice,
                encryptedSecret = encryptedDeviceSecret
            )

            actualDeviceSecret shouldBe (deviceSecret)
        }

        "not render its content via toString" {
            keyStoreEncryptionSecret.toString() shouldBe ("Secret(${keyStoreEncryptionSecret.javaClass.name})")
        }
    }
})
